### PR TITLE
Add automatic redirect creation V2

### PR DESCRIPTION
**Pull request by ababic**
(re)implements the 'backend' requirements for https://github.com/wagtail/rfcs/pull/34

Rejigged slightly from the changes in #7761 / #7774 to make the handler enabled by default, and added a test to show the effect of deactivating.

### CODE REVIEW

**Review by kaedroho — COMMENTED** [review]
Thanks for resubmitting this @ababic! I've pointed out a few things that I should be fixed and opened some discussions.

Things to fix:
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777037114
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777041897
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777043351

Discussions: (ping @gasman)
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777037114
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777042510
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777142180
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777041012
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777140895

File: CHANGELOG.txt [hunk 1]
```diff
@@ -15,6 +15,9 @@ Changelog
  * Fix: Rename additional 'spin' CSS animations to avoid clashes with other libraries (Kevin Gutiérrez)
  * Fix: `default_app_config` deprecations for Django >= 3.2 (Tibor Leupold)
  * Removed WOFF fonts
+ * Add `page_url_path_changed` signal for pages (Andy Babic)
+ * Fix: 'Page move' actions being incorrectly logged as 'Page reorder' under some circumstances (Andy Babic)
+ * Add automatic redirect creation (Andy Babic, with sponsorship from [The National Archives](https://www.nationalarchives.gov.uk))
```
Comment (kaedroho) [inline CHANGELOG.txt:20]: We should move this up to the top with the other new features

File: docs/reference/contrib/redirects.rst [hunk 2]
```diff
@@ -40,6 +40,23 @@ Page model recipe of to have redirects created automatically when changing a pag
 
 For an editor's guide to the interface, see :ref:`managing_redirects`.
 
+Automatic redirect creation
+===========================
+
+.. versionadded:: 2.16
+
+By default, Wagtail automatically create permanent redirects for pages (and their descendants) when they are moved or have their slug changed. This helps to preserve SEO rankings of pages over time, and helps site visitors get to the right place when using bookmarks or using outdated links.
```
Comment (kaedroho) [inline docs/reference/contrib/redirects.rst:48]:
```suggestion
Wagtail automatically creates permanent redirects for pages (and their descendants) when they are moved or their slug is changed. This helps to preserve SEO rankings of pages over time, and helps site visitors get to the right place when using bookmarks or using outdated links.
```

File: docs/reference/contrib/redirects.rst [hunk 3]
```diff
@@ -40,6 +40,23 @@ Page model recipe of to have redirects created automatically when changing a pag
...
+Disabling automatic redirect creation
+-------------------------------------
+
+Wagtail's default implementation works best for small-to-medium sized projects (5000 pages or fewer) that mostly use Wagtail's built-in methods for URL generation (Overrides to `Page.get_url_parts()` will be respected, but require additional database queries for each item where custom model fields are used to generate the return value).
+
+If your project does a lot of overriding of URL-related methods, or has sections with tens (or hundreds) of thousands of pages that you want to be able to move freely, you may want to disable this feature by adding the following to your project settings:
+
+.. code-block:: python
+
+  WAGTAILREDIRECTS_AUTOCREATE = False
```
Comment (kaedroho) [inline docs/reference/contrib/redirects.rst:59]: Do we need an underscore between `AUTO` and `CREATE`?

File: wagtail/contrib/redirects/models.py [hunk 4]
```diff
@@ -27,6 +29,25 @@ class Redirect(models.Model):
         on_delete=models.CASCADE
     )
     redirect_link = models.URLField(verbose_name=_("redirect to any URL"), blank=True, max_length=255)
+    automatically_created = models.BooleanField(
+        verbose_name=_("automatically created"),
+        default=False,
+        editable=False,
+    )
+    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now=True, null=True)
```
Comment (kaedroho) [inline wagtail/contrib/redirects/models.py:37]: Fix: ``auto_now`` should be ``auto_now_add``

File: wagtail/contrib/redirects/models.py [hunk 5]
```diff
@@ -27,6 +29,25 @@ class Redirect(models.Model):
...
+        blank=True,
+        null=True,
+        related_name='+',
+    )
+    trigger_id = models.PositiveIntegerField(
+        blank=True,
+        null=True,
+        db_index=True,
+    )
+    trigger = GenericForeignKey("trigger_content_type", "trigger_id")
```
Comment (kaedroho) [inline wagtail/contrib/redirects/models.py:50]: It looks like this is a foreign key to the action log that created the redirect. It doesn't appear to get used anywhere though. Do we need to capture this?

(If we're not entirely sure we need this, I think we should leave it out and add it back in later if the need arises)

File: wagtail/contrib/redirects/signal_handlers.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+) -> None:
+    if (
+        not getattr(settings, "WAGTAILREDIRECTS_AUTOCREATE", True)
+        or instance.is_site_root()
+    ):
+        return None
+
+    logger.info(f"Creating redirects for page: '{instance}' id={instance.id}")
+    batch = BatchRedirectCreator(max_size=2000, ignore_conflicts=True)
+    request = get_dummy_request()
```
Comment (kaedroho) [inline wagtail/contrib/redirects/signal_handlers.py:54]: Would it be possible to implement this optimisation without creating a dummy request? Looks like we only need an object to hold the site route paths so they can be reused between calls, so there must be a simpler way to do that.

Also, the redirect object is passed in to `Site.find_for_request` but the dummy request would always have the same domain regardless of the page. Would this work when generating redirects on non-default sites?

Do you think it would be possible to pass the site to ``Page.get_url_paths`` instead of the request, and bypass the logic that finds the site? (ping @gasman)

File: wagtail/contrib/redirects/signal_handlers.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            # Use modified page to retrieve old site and url
+            old_site_id, _, old_url = page.get_url_parts(request)
+        except TypeError:
+            continue
+
+        if (old_site_id, old_url) != (new_site_id, new_url):
+            page_paths = [old_url]
+            page_paths.extend(
+                old_url.rstrip("/") + path
+                for path in page.get_cached_paths()
```
Comment (kaedroho) [inline wagtail/contrib/redirects/signal_handlers.py:78]: ``get_cached_paths`` is intended to be used for cache invalidation, it may include variants of GET parameters which wouldn't be very useful for redirects as they only look at the path.

Could we add a separate method to allow specifying the paths for creating redirects? This would allow developers to provide a different set of URLs if they need to.

Alternatively, I think we should at least think of a new name for ``get_cached_paths`` if we're going to start using it for non-cache related things.

File: wagtail/core/models/__init__.py [hunk 8]
```diff
@@ -1547,6 +1561,16 @@ def move(self, target, pos=None, user=None):
         )
         logger.info("Page moved: \"%s\" id=%d path=%s", self.title, self.id, new_url_path)
 
+        if url_path_changed:
```
Comment (kaedroho) [inline wagtail/core/models/__init__.py:1564]: All the existing methods on page hide away the ``url_path`` and express page URLs with a site/path combination. So I wonder if we should implement this signal handler to be site-aware? I think this will simplify the logic in ``autocreate_redirects`` function too (no need to hack the ``url_path`` field back to its old value to find the old path, since the old path itself would be to hand).

Do you think we could rename this to ``page_url_changed`` and replace the ``url_paths`` with ``site/path`` instead? If the page is visible on multiple sites, we could call this signal handler multiple times.

So, for example, instead of:

```python
page_url_path_changed(instance=page, url_path_before="/home/about/", url_path_after="/home/renamed/")
```

Could we do:

```python
page_url_changed(instance=page, site_before=site, path_before="/about/", site_after=site, path_after="/renamed/")
```

**Review by kaedroho — CHANGES_REQUESTED** [review]
Thanks for resubmitting this @ababic! I've pointed out a few things that I should be fixed and opened some discussions.

Things to fix:
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777036981
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777041897
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777043351

Discussions: (ping @gasman)
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777037114
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777042510
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777142180
- https://github.com/wagtail/wagtail/pull/7809#discussion_r777041012
 - https://github.com/wagtail/wagtail/pull/7809#discussion_r777140895

File: docs/reference/contrib/redirects.rst (hunk 3 above)
Comment (ababic) [reply docs/reference/contrib/redirects.rst:59]: Unsure on this one. I was going for something descriptive and compact. I certainly don't mind adding an underscore. Googling around, 'autocreation' appears to be a well-enough known term. 'autocreate' less so, though.

File: wagtail/contrib/redirects/models.py (hunk 4 above)
Comment (ababic) [reply wagtail/contrib/redirects/models.py:37]: @kaedroho Good catch! Do you know off the top of your head if redirect changes are covered by logging? I'm wondering what should happen if an automatically created redirect is manually updated? Should they be unmarked as automatically created at that point? Should there be a checkbox on the edit form perhaps?

File: wagtail/contrib/redirects/models.py (hunk 5 above)
Comment (ababic) [reply wagtail/contrib/redirects/models.py:50]: Yes, the feature really needs a bit of UX input to figure out where best to surface this. In the meantime though, I think logging this while we can is still a good idea... It means that if/when we do surface it, the data is more likely to be available. The data is also available for debugging, should a developer need to dig into it. What do you think?

File: wagtail/contrib/redirects/signal_handlers.py (hunk 6 above)
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:54]: I see what you mean. Truth is, I find the various URL methods a little confusing to this day. I was under the impression that `request` could always be any old thing, but that isn't true for all of them. get_url_parts() does not care about the value type, but passing it a value called `request` that isn't a HttpRequest just to benefit from caching feels slightly more hacky to me. It's a strange API really. I would much rather url methods always receive a `HttpRequest`, and there be a lightweight helper (like this) to generate dummy ones when a real one isn't available. Or maybe a better route to consistency is to revise the URL methods so that `request` can *always* be any old value?

File: wagtail/contrib/redirects/signal_handlers.py (hunk 7 above)
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:78]: Thanks. Okay, I will think on this and try to come up with something sensible.

File: wagtail/core/models/__init__.py (hunk 8 above)
Comment (ababic) [reply wagtail/core/models/__init__.py:1564]: Thinking about where the complexity lives is definitely useful, but I feel this suggestion gets it wrong in a couple of ways:
1. It involves complicating the Page.save() and Page.move() methods further in order to have all of those values to hand, for functionality that is entirely optional. The current solution is lightweight in terms of changes to `Page`, because `page_url_path_changed()` uses values that are mostly already available.
2. It assumes a lot more about how URLs work within a project. 

While we should not complicate things unnecessarily for every edge-case we can think of, if having a little more complication in the (optional) signal handler means that we keep everything else simple/generic, I would say that's probably the better way to go?

Comment (kaedroho) [reply wagtail/core/models/__init__.py:1564]: > It involves complicating the Page.save() and Page.move() methods further in order to have all of those values to hand, for functionality that is entirely optional. The current solution is lightweight in terms of changes to Page, because page_url_path_changed() uses values that are mostly already available.

The current method probably makes sense for automatic redirects, but if this is meant to become part of the public API for pages, I feel it's important to be consistent with the existing methods.

I'd think we'd implement this as a method or a function so it wouldn't have to make the implementations of those functions any more complicated than they would be if they were using ``url_path``?

Alternatively, could we implement a signal for changing the slug instead? And use the ``page_moved`` signal?

Comment (ababic) [reply wagtail/core/models/__init__.py:1564]: > The current method probably makes sense for automatic redirects, but if this is meant to become part of the public API for pages, I feel it's important to be consistent with the existing methods.

Okay, I see the importance of that too. Maybe there's some meeting in the middle to be had - I'm just not seeing it yet.

> Alternatively, could we implement a signal for changing the slug instead? And use the `page_moved` signal?

I'd be happy to go with that, but the `url_path_before` and `url_path_after` values are really crucial to the current signal handler solution, and I'm not sure they would be a good fit for a `page_slug_changed` signal.

If we can figure out an alternative way to retrieve all of the old page URLs without the `url_path_before` and `url_path_after` values (and the current in-memory change trick), I'd be more than happy using `page_slug_changed` and `post_page_move`.

File: wagtail/contrib/redirects/signal_handlers.py (hunk 6 above)
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:54]: >  get_url_parts() does NOT care about the value type

Actually, I have this wrong. Unless `request` is `None` or has both `get_host()` and `get_port()` methods, things break just beyond [this point](https://github.com/wagtail/wagtail/blob/main/wagtail/core/models/__init__.py#L1171) - so, using a `HttpRequest` is really the best option as far as I can see.

Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:54]: > Do you think it would be possible to pass the site to Page.get_url_paths

Ah, you mean `get_url_parts` here.. the typo threw me. I'm with you now.

Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:54]: > The dummy request would always have the same domain regardless of the page. Would this work when generating redirects on non-default sites?

Provided the page only belongs to a single site, it should be fine. I suppose we could instead determine which sites the page belongs to in advance, then repeatedly call `get_url_parts` for each site.

File: wagtail/contrib/redirects/signal_handlers.py (hunk 7 above)
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py]: @kaedroho I have added a new `get_route_paths()` method for this now. Open to suggestions on the name :)

File: wagtail/core/models/__init__.py (hunk 8 above)
Comment (ababic) [reply wagtail/core/models/__init__.py]: @kaedroho  `page_url_path_changed` is now gone, and in it's place is:

```
page_slug_changed(instance=page, instance_before=old_page)
```

We're not claiming that the url has changed, so isn't misleading. And it's useful to have the old and new versions to hand, because it allows you to see ALL of the fields that changed (not just the `slug`).

File: wagtail/contrib/redirects/models.py (hunk 5 above)
Comment (ababic) [reply wagtail/contrib/redirects/models.py]: In the spirit of keeping things simple, I've removed the capturing of the log action.

File: wagtail/contrib/redirects/signal_handlers.py (hunk 6 above)
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py]: Am open to suggestions on the name, but I've added [`Page.get_url_parts_for_all_sites()`](https://github.com/wagtail/wagtail/blob/774b82f97feacd3d7c2f635b7328b5b3b84aa3d8/wagtail/core/models/__init__.py#L1310-L1325) to help with this.

This and [`Page.get_relevant_site_root_paths`](https://github.com/wagtail/wagtail/blob/774b82f97feacd3d7c2f635b7328b5b3b84aa3d8/wagtail/core/models/__init__.py#L1190-L1204)  allow us to get the values we need without changing the signature or `get_url_parts()` and without using the dummy request.

Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py]: I realise now that a method separate to `get_url_parts()` will not work, as we need to pick up existing overrides for that method. So, I have added a `site` kwarg to `get_url_parts()` after all. The method was getting rather unwieldily though, so have broken it up a little to help with readability

**Comment by ababic** [comment]
@kaedroho / @gasman This PR is now ready for a second review, having implemented all of the previous feedback. I apologise in advance for the sprawl... I probably could have gotten away without breaking up `Page.get_url_parts()` and `Site.get_site_root_paths()` quite so much, but I I hope you'll agree it makes the code easier to digest? On the whole, I think the changes help to dispel some of the 'magic' around how URLs work, and should make it easier for devs.

File: wagtail/core/models/__init__.py [hunk 9]
```diff
@@ -497,21 +507,27 @@ def save(self, clean=True, user=None, log_action=False, **kwargs):
...
+                    slug_changed = True
                     old_url_path = old_record.url_path
                     new_url_path = self.url_path
 
         result = super().save(**kwargs)
 
-        if not is_new and update_descendant_url_paths:
+        if slug_changed:
             self._update_descendant_url_paths(old_url_path, new_url_path)
+            # Emit page_slug_changed signal on successful db commit
```
Comment (kaedroho) [inline wagtail/core/models/__init__.py:521]: Just wondering why this needs to be deferred to after the transaction?

This behaviour might not be obvious to a developer who is using this signal, so they might think that their logic is running in the transaction when it's not.

File: wagtail/core/models/__init__.py [hunk 10]
```diff
@@ -1119,44 +1135,109 @@ def _get_site_root_paths(self, request=None):
...
-        if site:
-            for site_id, root_path, root_url, language_code in possible_sites:
-                if site_id == site.pk:
-                    break
-            else:
-                site_id, root_path, root_url, language_code = possible_sites[0]
+        # fallback to the first item
+        return options[0]
 
+    def get_url_parts(self, request=None, site=None):
```
Comment (kaedroho) [inline wagtail/core/models/__init__.py:1189]: This refactoring of core routing logic is quite extensive. I think it would be best to do this in a separate PR so it could be more carefully considered

File: wagtail/core/models/__init__.py (hunk 9 above)
Comment (ababic) [reply wagtail/core/models/__init__.py:521]: Good question. I was being cautious here really... I see the sort of things you implement with signals as 'side effects' that aren't part of the same transaction, and an issue in a 'side effect' should prevent the core functionality from working. Issuing signals in a transaction also means that implementers are more restricted. For example, you can't reliably go and do things in a new thread until the database changes are committed.

File: wagtail/core/models/__init__.py (hunk 10 above)
Comment (ababic) [reply wagtail/core/models/__init__.py:1189]: With all due respect, it's been a chore getting this far, and I am realistically out of time. I would very much appreciate it if we could do the considering here, with the changes all together?

**Comment by ababic** [comment]
Closing in favour of #7827
